File: src/main/java/coza/trojanc/receipt/template/process/impl/DefaultTemplateProcessor.java
package coza.trojanc.receipt.template.process.impl;

import com.sun.deploy.util.StringUtils;
import coza.trojanc.receipt.context.ContextMap;
import coza.trojanc.receipt.context.impl.DefaultContextMap;
import coza.trojanc.receipt.template.PrintTemplate;
import coza.trojanc.receipt.template.fields.*;
import coza.trojanc.receipt.template.process.ProcessedTemplate;
import coza.trojanc.receipt.template.process.TemplateProcessor;
import coza.trojanc.receipt.template.process.fields.ProcessedFeed;
import coza.trojanc.receipt.template.process.fields.ProcessedFillLine;
import coza.trojanc.receipt.template.process.fields.ProcessedLine;
import coza.trojanc.receipt.template.process.fields.ProcessedText;

import static coza.trojanc.receipt.context.ContextResolver.ARRAY_LENGTH_SUFFIX;

/**
 * Created by Charl-PC on 2016-10-11.
 */
public class DefaultTemplateProcessor implements TemplateProcessor {

	/**
	 * The resulting processed template.
	 */
	private ProcessedTemplate processedTemplate;

	/**
	 * Context being used to retrieve dynamic values for the template.
	 */
	private ContextMap context;

	/**
	 * Process a template item.
	 * @param item The template item to process.
	 */
	private void processTemplateItem(TemplateLine item){
		this.processTemplateItem(item, null);
	}

	private void processTemplateItem(TemplateLine item, String repeatPrefix){

		// If it is a feed
		if(Feed.class.isAssignableFrom(item.getClass())){
			Feed feed = (Feed)item;
			for(int i = 0 ; i < feed.getFeedAmount(); i++){
				processedTemplate.getItems().add(new ProcessedFeed());
			}
		}

		else if(Line.class.isAssignableFrom(item.getClass())){
			processLine((Line)item, repeatPrefix);
		}

		else if(FillLine.class.isAssignableFrom(item.getClass())){
			processFillLine((FillLine)item);
		}

		else if(RepeatBlock.class.isAssignableFrom(item.getClass())){
			processRepeatBlock((RepeatBlock)item);
		}

	}


	/**
	 * Process a repeat block
	 * @param repeatBlock The repeat block to process
	 */
	private void processRepeatBlock(RepeatBlock repeatBlock){
		final String keyPrefix = repeatBlock.getRepeatOn();
		final int repeatSize = Integer.parseInt(context.get(keyPrefix+ARRAY_LENGTH_SUFFIX));
		for(int idx = 0 ; idx < repeatSize; idx++) {
			int iteration = idx;
			repeatBlock.getLines().forEach(line -> processTemplateItem(line, keyPrefix + "[" + iteration + "]"));
		}
	}

	/**
	 * Process a fill line.
	 * @param fillLine
	 */
	private void processFillLine(FillLine fillLine){
		processedTemplate.getItems().add(new ProcessedFillLine(fillLine.getCharacter()));
	}

	/**
	 * Process a line that can contain text
	 * @param line The line to process
	 */
	private void processLine(Line line, String repeatPrefix){
		ProcessedLine processedLine = new ProcessedLine();
		processedTemplate.getItems().add(processedLine);

		line.getLineItems().forEach(lineItem -> {
			if (Text.class.isAssignableFrom(lineItem.getClass())) {
				addStaticText((Text) lineItem, processedLine);
			}
			else if (DynamicText.class.isAssignableFrom(lineItem.getClass())) {
				addDynamicText((DynamicText) lineItem, processedLine, repeatPrefix);
			}
		});
	}

	private void addStaticText(Text text, ProcessedLine processedLine){
		ProcessedText processedText = new ProcessedText();
		processedText.setAlignment(text.getAlignment());
		processedText.setMode(text.getMode());
		processedText.setText(text.getText());
		processedText.setOffset(text.getOffset());
		processedLine.getLineItems().add(processedText);
	}

	private void addDynamicText(DynamicText text, ProcessedLine processedLine, String repeatPrefix){
		ProcessedText processedText = new ProcessedText();
		processedText.setAlignment(text.getAlignment());
		processedText.setMode(text.getMode());
		processedText.setOffset(text.getOffset());

		String contextKey;
		if(repeatPrefix == null){
			contextKey = text.getContextKey();
		}
		else{
			contextKey = repeatPrefix;
			if(!(text.getContextKey() == null || text.getContextKey().length() == 0)){
				contextKey += text.getContextKey();
			}
		}
		processedText.setText(context.get(contextKey));
		processedLine.getLineItems().add(processedText);
	}


	@Override
	public ProcessedTemplate process(PrintTemplate template) {
		return process(template, new DefaultContextMap(0));
	}

	@Override
	public ProcessedTemplate process(PrintTemplate template, ContextMap context) {
		this.processedTemplate = new ProcessedTemplate();
		this.context = context;

		template.getLines().forEach(this::processTemplateItem);
		return processedTemplate;
	}
}

File: src/main/java/coza/trojanc/receipt/template/process/impl/DefaultTemplateProcessor.kt
package coza.trojanc.receipt.template.process.impl

import coza.trojanc.receipt.context.ContextMap
import coza.trojanc.receipt.context.impl.DefaultContextMap
import coza.trojanc.receipt.template.PrintTemplate
import coza.trojanc.receipt.template.fields.*
import coza.trojanc.receipt.template.process.ProcessedTemplate
import coza.trojanc.receipt.template.process.TemplateProcessor
import coza.trojanc.receipt.template.process.fields.ProcessedFeed
import coza.trojanc.receipt.template.process.fields.ProcessedFillLine
import coza.trojanc.receipt.template.process.fields.ProcessedLine
import coza.trojanc.receipt.template.process.fields.ProcessedText

import coza.trojanc.receipt.context.ContextResolver.ARRAY_LENGTH_SUFFIX

/**
 * @author Charl Thiem
 */
class DefaultTemplateProcessor : TemplateProcessor {

    /**
     * The resulting processed template.
     */
    private var processedTemplate: ProcessedTemplate? = null

    /**
     * Context being used to retrieve dynamic values for the template.
     */
    private var context: ContextMap? = null

    /**
     * Process a template item.
     * @param item The template item to process.
     */
    private fun processTemplateItem(item: TemplateLine) {
        this.processTemplateItem(item, null)
    }

    private fun processTemplateItem(item: TemplateLine, repeatPrefix: String?) {

        // If it is a feed
        if (Feed::class.java.isAssignableFrom(item.javaClass)) {
            val feed = item as Feed
            for (i in 0..feed.feedAmount - 1) {
                processedTemplate!!.items.add(ProcessedFeed())
            }
        } else if (Line::class.java.isAssignableFrom(item.javaClass)) {
            processLine(item as Line, repeatPrefix)
        } else if (FillLine::class.java.isAssignableFrom(item.javaClass)) {
            processFillLine(item as FillLine)
        } else if (RepeatBlock::class.java.isAssignableFrom(item.javaClass)) {
            processRepeatBlock(item as RepeatBlock)
        }

    }


    /**
     * Process a repeat block
     * @param repeatBlock The repeat block to process
     */
    private fun processRepeatBlock(repeatBlock: RepeatBlock) {
        val keyPrefix = repeatBlock.repeatOn
        val repeatSize = Integer.parseInt(context!!.get(keyPrefix!! + ARRAY_LENGTH_SUFFIX))
        for (idx in 0..repeatSize - 1) {
            repeatBlock.lines.forEach { line -> processTemplateItem(line, "$keyPrefix[$idx]") }
        }
    }

    /**
     * Process a fill line.
     * @param fillLine
     */
    private fun processFillLine(fillLine: FillLine) {
        processedTemplate!!.items.add(ProcessedFillLine(fillLine.character))
    }

    /**
     * Process a line that can contain text
     * @param line The line to process
     */
    private fun processLine(line: Line, repeatPrefix: String?) {
        val processedLine = ProcessedLine()
        processedTemplate!!.items.add(processedLine)

        line.lineItems.forEach { lineItem ->
            if (Text::class.java.isAssignableFrom(lineItem.javaClass)) {
                addStaticText(lineItem as Text, processedLine)
            } else if (DynamicText::class.java.isAssignableFrom(lineItem.javaClass)) {
                addDynamicText(lineItem as DynamicText, processedLine, repeatPrefix)
            }
        }
    }

    private fun addStaticText(text: Text, processedLine: ProcessedLine) {
        val processedText = ProcessedText()
        processedText.alignment = text.alignment
        processedText.mode = text.mode
        processedText.text = text.text
        processedText.offset = text.offset
        processedLine.lineItems.add(processedText)
    }

    private fun addDynamicText(text: DynamicText, processedLine: ProcessedLine, repeatPrefix: String?) {
        val processedText = ProcessedText()
        processedText.alignment = text.alignment
        processedText.mode = text.mode
        processedText.offset = text.offset

        var contextKey: String?
        if (repeatPrefix == null) {
            contextKey = text.contextKey
        } else {
            contextKey = repeatPrefix
            if (!(text.contextKey == null || text.contextKey!!.length == 0)) {
                contextKey += text.contextKey
            }
        }
        processedText.text = context!!.get(contextKey)
        processedLine.lineItems.add(processedText)
    }


    override fun process(template: PrintTemplate): ProcessedTemplate {
        return process(template, DefaultContextMap(0))
    }

    override fun process(template: PrintTemplate, context: ContextMap): ProcessedTemplate {
        this.processedTemplate = ProcessedTemplate()
        this.context = context

        template.lines.forEach(Consumer<TemplateLine> { this.processTemplateItem(it) })
        return processedTemplate
    }
}
